[Baremetal] Introduce baremetal config and scripts to measure memory usage

hanno-arm opened the pull request:
__Summary:__ This PR introduces the test configuration file `configs/baremetal.h` which can be used to test minimal configurations enabling a single ciphersuite. Currently, the configuration targets ECDHE-ECDSA-AES-CCM-8. Further, it adds a modification `configs/baremetal_test.h` to `configs/baremetal.h`  which modifies `baremetal.h` in a minimal way to allow building and running tests and example programs (currently, numerous additions are necessary, but PRs like https://github.com/ARMmbed/mbedtls/pull/2260 will hopefully remove some of them). Finally, it adds a test script `scripts/baremetal.sh` which supports (a) creating a 32-bit Cortex-M build with `baremetal.h` and measuring the static memory usage of the resulting binary, and (b) creating a 32-bit host build with `baremetal.h + baremetal_test.h`, and measuring the dynamic memory usage in an example `ssl_client2 + ssl_server2` run.

This PR is still work in progress.

sbutcher-arm commented on configs/baremetal.h:
```diff
@@ -0,0 +1,113 @@
+ * \file baremetal.h
+ *
+ * \brief Test configuration for minimal baremetal Mbed TLS builds
+ *        based on the following primitives:
```
Specifically, we want to support only 
`TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8` as specified in the LWM2M spec [here](http://openmobilealliance.org/RELEASE/LightweightM2M/V1_1-20180612-C/OMA-TS-LightweightM2M_Transport-V1_1-20180612-C.pdf). I think you can state that to make it clearer.

hanno-arm replied:
I corrected the ciphersuite.

sbutcher-arm commented on configs/baremetal.h:
```diff
@@ -0,0 +1,113 @@
+#define MBEDTLS_BAREMETAL_CONFIG_H
+
+#define MBEDTLS_HAVE_TIME
+#define MBEDTLS_HAVE_TIME_DATE
```
I don't think we should define `MBEDTLS_HAVE_TIME_DATE`.

hanno-arm replied:
We need it for X.509 CRT verification.

sbutcher-arm commented on scripts/baremetal.sh:
```diff
@@ -0,0 +1,351 @@
+: ${BAREMETAL_USER_CONFIG:=./configs/baremetal_test.h}
+CFLAGS_USER_CONFIG="-DMBEDTLS_USER_CONFIG_FILE='\"../$BAREMETAL_USER_CONFIG\"'"
+
+: ${ARMC5_BIN_DIR:=""}
```
We only need ARMC6, but I think you might as well leave this in.

hanno-arm replied:
Ok, I leave it.

sbutcher-arm replied:
My mistake. We only need ARMCC5.

sbutcher-arm commented:
@mpg / @jarlamsa - Please review so I can merge this PR.

mpg commented on scripts/baremetal.sh:
```diff
@@ -0,0 +1,351 @@
+    VALGRIND_BASE="$VALGRIND --tool=massif $VALGRIND_MASSIF_PARAMS"
+
+    FUNC_IGNORE=""
+    FUNC_IGNORE="mbedtls_ssl_setup           $FUNC_IGNORE"
```
I don't think this should be ignored. Otherwise we'll not be able to measure improvements to how I/O buffers are managed.

hanno-arm replied:
Done.

mpg commented on scripts/baremetal.sh:
```diff
@@ -0,0 +1,351 @@
+
+    FUNC_IGNORE=""
+    FUNC_IGNORE="mbedtls_ssl_setup           $FUNC_IGNORE"
+    FUNC_IGNORE="mbedtls_mpi_grow            $FUNC_IGNORE"
```
I don't think this should be ignored. Otherwise we'll not be able to measure improvements switching to new ECC code.

hanno-arm replied:
Done.

mpg commented on scripts/baremetal.sh:
```diff
@@ -0,0 +1,351 @@
+    : ${VALGRIND:=valgrind}
+    : ${VALGRIND_MASSIF_PARAMS="--time-unit=B --threshold=0.01 --detailed-freq=1"}
+
+    if [ $1 -eq 1 ]; then
```
I think this would deserve a comment at the top of the function: "Usage: baremetal_ram_heap 0 does this and baremetal_ram_heap 1 does that."

hanno-arm replied:
Agreed.

hanno-arm replied:
Done.

mpg commented on scripts/generate_codesize_stats.sh:
```diff
@@ -0,0 +1,123 @@
+#!/bin/sh
+# generate_codesize_stats.sh
```
Some name nitpicking: for now all `scripts/generate_xxx` scripts generate something that's part of the library, and this is different - should we call it `format_codesize_stats.sh` or `extract_codesize_stats.sh` instead? ("Extract" is what the usage message says.)

hanno-arm replied:
Agreed - renaming to `extract_codesize_stats.sh`

mpg commented:
The CI wants us to know that `check-files.py` is unhappy about the presence of the string "TODO" in `configs/baremetal.h` on line 75.

mpg commented:
When using the script locally, I'm getting surprising results from massif:
```
% scripts/baremetal.sh --ram --heap | grep '^Max'                                                                                       git|remotes/hanno/baremetal_config|
Max heap usage: 7112 (heap 6653+459, stack 0)
Max heap usage: 31320 (heap 5779+357, stack 25184)
```

I somehow have trouble believing that most of our RAM usage happens on the stack. Unfortunately right now I don't understand callgrind's output enough to check is massif's stack figures are consistent with callgrind's.

Also, I find it surprising that out heap usage would appear to be lower when our stack usage is measured, but perhaps that's just a side-effect of different sampling?

Can anybody reproduce those surprising results? Or even understand them?

hanno-arm commented:
@mpg That might be due to the 20Kb stack buffer for the HTTP request in the `main()` of `ssl_client2.c`.

mpg commented:
@hanno-arm Oh, good point, that's a relief :) I think we should find a way to exclude it so that we have a correct idea what our stack usage actually is. Perhaps the most obvious way is to allocate that buffer on the heap instead, using a dedicated function so that we can easily exclude it from massif's measurements.

(I though for a moment that we might want instead to exclude all of `main()`s stack usage, but I'm not sure if that would be a fair thing to do: part of this stack usage comes from structures such as `mbedtls_ssl_config` and `mbedtls_ssl_context`, which we want measured.)

mpg commented:
I've spent some time readind the documentation of callgrind, but I couldn't easily figure out how to get meaningful stack usage figures from the `ram_stack_` output file. I think some documentation or even a helper script about that would be a useful addition to this PR (or a follow-up PR).

hanno-arm commented:
@mpg Another point is that currently `ssl_client2` allocates CRTs, SSL configuration, session and context on the stack - I wonder if we should also allocate them on the heap in the `baremetal` branch. WDYT?

mpg commented:
I don't know. I don't think it's unreasonable to have them on the stack, but perhaps we should have a way to measure what is the stack usage of `ssl_client2`'s `main()` function so that we can distinguish between (1) stack usage of Mbed TLS functions and (2) size of necessary variable in a reasonable application, that could live on the heap or stack depending in the needs of the application.

Perhaps this could be achieved by invoking `ssl_client2` with an invalid option and measuring its stack usage in that run.

hanno-arm commented on configs/baremetal_test.h:
```diff
@@ -0,0 +1,57 @@
+#define MBEDTLS_PEM_PARSE_C
+#define MBEDTLS_BASE64_C
+/* We don't have Secp256r1 test CRTs at the moment. */
+#define MBEDTLS_ECP_DP_SECP384R1_ENABLED
```
Yes, thanks!

hanno-arm commented on configs/baremetal.h:
```diff
@@ -0,0 +1,113 @@
+#define MBEDTLS_SSL_EXTENDED_MASTER_SECRET
+#define MBEDTLS_SSL_SESSION_TICKETS
+#define MBEDTLS_SSL_COOKIE_C
+#define MBEDTLS_SSL_PROTO_DTLS /* TODO: Unset TLS one possible. */
```
Done.

hanno-arm commented on configs/baremetal.h:
```diff
@@ -0,0 +1,113 @@
+#define MBEDTLS_ECP_C
+#define MBEDTLS_ECP_DP_SECP256R1_ENABLED
+#define MBEDTLS_ECP_NIST_OPTIM
+#define MBEDTLS_ECDSA_DETERMINISTIC
```
Done.

hanno-arm commented:
@mpg Thank you for your review! I believe I have addressed your comments, could you please re-review?

@jarlamsa @sbutcher-arm Could you re-review, too, please?

In regards to measuring stack usage: Previously, I only ran `callgrind` but didn't do any processing on the generated output. As @mpg noted, that's not enough to measure stack usage. As a remedy, I have now added `-fstack-usage` to `CFLAGS` to measure per-function stack usage, and also added a run to `cflow` which tracks function calls _statically_, by analyzing the source code. 

The generates files should be enough to study stack usage on a per-function basis, but there is still some scripting to be done for that, which I'd suggest we move outside of this PR.

mpg commented on scripts/baremetal.sh:
```diff
@@ -263,12 +263,23 @@ baremetal_ram_stack() {
     echo "Done"
 
+    # Extract callgraphs from source files directly
+    if [ -x "$(command -v cflow)" ]; then
```
Oh, I didn't know about `command`, I usually use `which` for that, but apparently `which` is not specified by POSIX, while `command` is, so thanks for letting me know about it :)

mpg commented:
Note: I'm done a bit of research because I'm quite surprised that there isn't a ready-made tool for combining the output of GCC's `-fstack-usage` with the output of `cflow` into nice readable stack figures, but apparently there isn't, and interested users are supposed to write their own script. I've however noticed two scripts that could be of interest, at least as sources of inspiration for writing our own: https://github.com/PeterMcKinnis/WorstCaseStack (recent, python3), https://dlbeer.co.nz/oss/avstack.html (older, Perl) - so I'm mentioning them here for reference.

mpg commented:
ping @jarlamsa for re-review after the latest changes
ping @sbutcher-arm for re-review and hopefully merging if you're still happy with it

sbutcher-arm commented:
Approved by myself. CI is passing (with exception of those pesky BSD timing tests that always fail). Just needs one more review approval (as @hanno-arm had to do some rework).

Ping @mpg, @jarlamsa.

mpg commented:
Reviewed and re-approved. Now has two approvals. Ping @sbutcher-arm
